### ISSUE TITLE
Investigate ways to simplify/improve the CI pipeline setup

### ISSUE THREAD

**Issue by nrfox** [kiali/kiali#7396]
### What do you want to improve?

There are many different scripts involved in setting up the Kiali CI (and local) testing environments which have evolved organically over time. Maintaining these is becoming more of maintenance burden. This issue is to investigate if there are ways that we can make them simpler and/or better.

**Comment by nrfox** [comment]
> I'm not sure if many on the Kiali dev team are proficient in python either.

And that's an important thing to consider before embarking on rewriting anything. script looks promising since it's written in go.

**Comment by leandroberetta** [comment]
I think I prefer moving to Python or Go, I'm fine with bash but I'm not so use to write it and sometimes I feel lost when something breaks. Having said that, I think that bash will always work out of the box after cloning the repo, probably if we move to Go or Python, the user wil need to install these tools too.

**Comment by jshaughn** [comment]
I think we have less Python expertise than we do for bash.  So if we need to make a change I'd prefer script or Go over Python.  Having said that, embarking on re-writes should probably be very targeted, how major of a problem do we really have here?

**Comment by josunect** [comment]
A simple improvement could be to use something like ShellCheck, a linter that find bugs, security flaws, and syntax errors in Bash, Sh, and Ksh scripts.
Maybe it could be used with a pre-commit hook.

**Comment by nrfox** [comment]
+1 using ShellCheck would be useful. ShellCheck alone doesn't go far enough in easing the maintenance burden though since it won't help with:
1. Passing structured data around to different shell scripts
2. Parallel execution
3. Compiler errors when changing variables across files
4. Reusing code (either internally or with libraries)

Once shell scripts grow to a certain size (more than one?) they become hard to maintain since they struggle with all the above.

**Comment by josunect** [comment]
> +1 using ShellCheck would be useful. ShellCheck alone doesn't go far enough in easing the maintenance burden though since it won't help with:
> 
>     1. Passing structured data around to different shell scripts
> 
>     2. Parallel execution
> 
>     3. Compiler errors when changing variables across files
> 
>     4. Reusing code (either internally or with libraries)
> 
> 
> Once shell scripts grow to a certain size (more than one?) they become hard to maintain since they struggle with all the above.

Yes, right. I was thinking as a way to help reducing some errors on what we already have, while other solutions are not implemented.

**Comment by ScriptingShrimp** [comment]
+1 to shell check, ai agents will have valuable feedback from that

**Comment by jshaughn** [comment]
I will take this one to help move it forward or close it.

**Comment by jshaughn** [comment]
OK, this is a tough one because there are definitely advantages to the Go approach. The type-safety and "silent-failure" issues are definite cons to the current approach. Go's potential parallel execution could save time in job setup, especially for MC jobs (although I am a little concerned that these limited runners may actually have resource limit issues if trying to do parallel setup). But the real issue here is just that the scripts are working and portable to all of the test envs. The sheer size of the scripts is a con for maintenance but also a con for migration, which would be a giant effort. I'd say that when this was proposed it was better timing, as we could easily get buried in the complexity of the scripts, making their maintenance error-prone. But with AI-assist the script management is much easier. The same could be said for the migration effort, but that sideways effort just doesn't seem necessary.

So, We will not proceed with the Go approach at this time.  There are two things we can do to improve the script approach. One is to look at shell-check, the other is to more consistetently  use `-e`, `-o pipefail`, `-u` options.

We can close the go-based PRs and update this Epic to enhance the existing scripts.  We will not use Python as this team is not very Python-aware.
